### DISCUSSION TITLE
onChange(of:) not working for childState

### DISCUSSION THREAD

**Discussion by Eunno-An** [pointfreeco/swift-composable-architecture#2926] [General]
Hello?
I have a question..

This is a parent Feature code.
It use child feature 'ProfileHeaderFeature'
But, the BindingReducer Never called.
Why it happened?

*profileHeaderFeature inside, there is BindingReducer.
*So i tested another child feature not have BindingReducer.

```swift
@Reducer
struct ProfileFeature {
    @ObservableState
    struct State {
        var profileHeaderFeature: ProfileHeaderFeature.State?
        ...
        
        init(profileHeaderFeature: ProfileHeaderFeature.State, introFeature: IntroFeature.State, hashTagComponentFeatures: IdentifiedArrayOf<HashTagComponentFeature.State>, saveFeature: SaveFeature.State) {
            self.profileHeaderFeature = profileHeaderFeature
            ...
        }
    }
    enum Action: BindableAction {
        case binding(BindingAction<State>)
        case profileHeaderFeature(ProfileHeaderFeature.Action)
        ...
    }
    var body: some ReducerOf<Self> {
        BindingReducer()
            .onChange(of: \.introFeature) { oldValue, newValue in
                self._printChanges()
            }
        Scope(state: \.profileHeaderFeature!, action: \.profileHeaderFeature) {
            ProfileHeaderFeature()
        }
...
}
```

**Comment by stephencelis** [comment]
@Eunno-An We might need more info or a project to work with, but note that `onChange` only detects changes made by the reducer it's attached to, which means it'll only detect changes to `\.introFeature` by the `BindingReducer` installed in `ProfileFeature`. Also, note that re-embedding `self._printChanges()` inside the `onChange` is probably not what you want, as that will recursively re-run `self` _again_, which could lead to an infinite loop. If you want to simply debug in here, you could call `let _ = print("introFeature", oldValue, newValue)`.

To detect changes made by _all_ reducers in the body, you must wrap everything with `CombineReducers` and apply `onChange` to that:

```diff
 var body: some ReducerOf<Self> {
+  CombineReducers {
     BindingReducer()
-      .onChange(of: \.introFeature) { oldValue, newValue in
-        self._printChanges()
-      }
     Scope(state: \.profileHeaderFeature!, action: \.profileHeaderFeature) {
       ProfileHeaderFeature()
     }
     ...
+  }
+  .onChange(of: \.introFeature) { oldValue, newValue in
+    let _ = print("introFeature", oldValue, newValue)
+  }
 }
```
